README.md
<img src="https://github.com/AraBlocks/docs/blob/master/ara.png" width="30" height="30" /> ara-network-node-dht
====================

[![Build Status](https://travis-ci.com/AraBlocks/ara-network-node-dht.svg?token=r6p7pesHZ9MRJsVsrYFe&branch=master)](https://travis-ci.com/AraBlocks/ara-network-node-dht)

An Ara Network node that runs a DHT.

## Status
This project is still in alpha development.

> **Important**: While this project is under active development, run `npm link` in `ara-network-node-dht` directory & `npm link ara-network-node-dht` in `ara-network` directory.

## Dependencies
- [Node](https://nodejs.org/en/download/)
- [ara-network][ara-network]

## Installation
```sh
$ npm install ara-network ara-network-node-dht
```

## Configuration
[ara-runtime-configuration][ara-runtime-configuration] is a dependency of [ara-network][ara-network] and will either read from the nearest `.ararc`.  Install [ara-runtime-configuration][ara-runtime-configuration] separately to specify default values not present in an `.ararc`.

Runtime configuration can be specified by targeting the
`[network.node.dht]` _INI_ section or a nested _JSON_ object
`{ network: { node: { dht: { ... }}}}`.

### Examples
_INI_ format in an `.ararc`:
```ini
[network.node.dht]
  ; maximum query concurrency

  concurrency = 16
  ; maximum tables stored
  maxTables = 1000

  ; maximum values stored
  maxValues = 1000

  ; in milliseconds
  timeout = 2000

  ; how long records (peers) are cached
  maxAge = 0

  ; bootstrap nodes
  nodes[] = 192.168.0.111

  ; port for DHT server to listen on
  port = 6881

  ; number of buckets (k-buckets)
  k = 20
```
_JSON_ format (passed in to ara-runtime-configuration `rc` function:
```json
"network": {
  "node": {
    "dht": {
      "multicast": true,
      "loopback": true,
      "port": 6881
    }
  }
}
```

## Usage
```js
const { argv } = require('yargs')
const dht = require('ara-network-node-dht')
const rc = require('ara-runtime-configuration')

void async function main() {
  try { await dht.configure(rc.network.node.dht, require('yargs')) }
  catch (err) { await dht.configure(null, require('yargs')) }
  await dht.start(argv)
}()
```

#### Command Line (ann)
```bash
$ ann --type dht --help
usage: ann -t dht [options]

Options:
  --debug, -D  Enable debug output                                     [boolean]
  --conf, -C   Path to configuration file                               [string]
  --help, -h   Show help                                               [boolean]
  --port, -p   Port or ports to listen on               [number] [default: 6881]
```

### Examples
#### Command Line (ann)
Invoke a network node with the `ann` (or `ara-network-node`) command line interface:
```sh
$ ann --type dht
```
Default port is `6881`.

## Contributing
- [Commit message format](/.github/COMMIT_FORMAT.md)
- [Commit message examples](/.github/COMMIT_FORMAT_EXAMPLES.md)
- [How to contribute](/.github/CONTRIBUTING.md)

## See Also
- [ara-network](https://github.com/arablocks/ara-network)
- [bittorrent-dht](https://www.npmjs.com/package/bittorrent-dht)
- [k-rpc](https://github.com/mafintosh/k-rpc)
- [k-rpc-socket](https://github.com/mafintosh/k-rpc-socket)

## License
LGPL-3.0

[ara-network]: https://github.com/arablocks/ara-network
[ara-runtime-configuration]: https://github.com/arablocks/ara-runtime-configuration



CHANGELOG.md
<a name="0.3.1"></a>
## [0.3.1](https://github.com/AraBlocks/ara-network-node-dht/compare/0.3.0...0.3.1) (2018-10-18)



<a name="0.3.0"></a>
# [0.3.0](https://github.com/AraBlocks/ara-network-node-dht/compare/0.2.0...0.3.0) (2018-08-16)



<a name="0.2.0"></a>
# [0.2.0](https://github.com/AraBlocks/ara-network-node-dht/compare/0.1.1...0.2.0) (2018-08-16)


### Bug Fixes

* Fix `ports` to `port` typo ([9d40372](https://github.com/AraBlocks/ara-network-node-dht/commit/9d40372))


### Features

* **Dockerfile:** Adding Dockerfile ([043bebc](https://github.com/AraBlocks/ara-network-node-dht/commit/043bebc))
* **index.js:** adding arguement parser ([3b57f4a](https://github.com/AraBlocks/ara-network-node-dht/commit/3b57f4a))
* **Makefile:** Adding Makefile for docker builds ([d233749](https://github.com/AraBlocks/ara-network-node-dht/commit/d233749))



<a name="0.1.1"></a>
## [0.1.1](https://github.com/AraBlocks/ara-network-node-dht/compare/0.1.0...0.1.1) (2018-05-21)



<a name="0.1.0"></a>
# [0.1.0](https://github.com/AraBlocks/ara-network-node-dht/compare/c2ab0f1...0.1.0) (2018-05-21)


### Features

* **index.js:** Initial implementation ([c2ab0f1](https://github.com/AraBlocks/ara-network-node-dht/commit/c2ab0f1))





index.js
const { info, warn } = require('ara-console')
const extend = require('extend')
const debug = require('debug')('ara:network:node:dht')
const dht = require('ara-network/dht')
const rc = require('ara-network/rc')()

const conf = {
  concurrency: 16,
  maxTables: 1000,
  maxValues: 1000,
  // in milliseconds
  timeout: 2000,
  // in milliseconds
  maxAge: 0,
  nodes: [],
  port: 6881,
  // number of buckets (k-buckets)
  k: 20,
}

if (rc.network && rc.network.node && rc.network.node.dht) {
  extend(true, conf, rc.network.node.dht)
}

let server = null

async function getInstance() {
  return server
}

async function configure(opts, program) {
  if (opts) {
    conf.concurrency = opts.concurrency
    conf.timeout = opts.timeout
    conf.port = opts.port
  }

  if (program) {
    const { argv } = program
      .option('p', {
        alias: 'port',
        type: 'number',
        describe: 'Port or ports to listen on',
        default: conf.port
      })
      .option('c', {
        alias: 'concurrency',
        type: 'number',
        describe: 'Max concurrency',
        default: conf.concurrency
      })

    if (argv.port) {
      // eslint-disable-next-line no-param-reassign
      conf.port = argv.port
    }

    if (argv.timeout) {
      // eslint-disable-next-line no-param-reassign
      conf.timeout = argv.timeout
    }

    if (argv.concurrency) {
      // eslint-disable-next-line no-param-reassign
      conf.concurrency = argv.concurrency
    }
  }
}

async function start() {
  if (server) { return false }

  server = dht.createServer(conf)
  server.listen(conf.port)
  server.on('error', onerror)
  server.on('close', onclose)
  server.on('listening', onlistening)

  return true

  function onerror(err) {
    warn('dht: error:', err.message)
    debug('error:', err)
  }

  function onclose() {
    warn('dht: Closed')
  }

  function onlistening() {
    const { port } = server.address()
    info('dht: Listening on port %s', port)
  }
}

async function stop() {
  if (null == server) { return false }
  warn('dht: Stopping server')
  server.destroy(onclose)
  return true
  function onclose() {
    server = null
  }
}

module.exports = {
  getInstance,
  configure,
  start,
  stop,
}
